profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/features/DataCenterApi.java

```diff
@@ -0,0 +1,156 @@
+ */
+package org.apache.jclouds.profitbricks.rest.features;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import com.google.inject.Inject;
+import com.google.inject.TypeLiteral;
+import org.jclouds.http.filters.BasicAuthentication;
+import org.jclouds.rest.annotations.RequestFilters;
+import java.io.Closeable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Named;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import org.apache.jclouds.profitbricks.rest.domain.DataCenter;
+import org.apache.jclouds.profitbricks.rest.domain.options.DepthOptions;
+import org.apache.jclouds.profitbricks.rest.features.DataCenterApi.DataCenterParser;
+import org.jclouds.Fallbacks;
+import org.jclouds.Fallbacks.EmptyListOnNotFoundOr404;
+import org.jclouds.Fallbacks.NullOnNotFoundOr404;
+import org.jclouds.http.HttpRequest;
+import org.jclouds.http.functions.ParseJson;
+import org.jclouds.json.Json;
+import org.jclouds.rest.annotations.Fallback;
+import org.jclouds.rest.annotations.MapBinder;
+import org.jclouds.rest.annotations.PATCH;
+import org.jclouds.rest.annotations.PayloadParam;
+import org.jclouds.rest.annotations.ResponseParser;
+import org.jclouds.rest.annotations.SelectJson;
+import org.jclouds.rest.binders.BindToJsonPayload;
+
+@Path("/datacenters")
+@RequestFilters(BasicAuthentication.class)
+public interface DataCenterApi extends Closeable {
+
+   @Named("datacenter:list")
+   @GET
+   @Produces("application/vnd.profitbricks.resource+json")
+   @SelectJson("items")
+   @Fallback(EmptyListOnNotFoundOr404.class)
+   List<DataCenter> list();
+   
+   @Named("datacenter:list")
+   @GET
+   @Produces("application/vnd.profitbricks.resource+json")
+   @SelectJson("items")
+   @Fallback(EmptyListOnNotFoundOr404.class)
+   List<DataCenter> list(DepthOptions options);
+   
+   /**
+    * @param identifier Data Center identifier
+    * @return Returns information about an existing virtual data center's state and configuration or <code>null</code>
+    * if it doesn't exist.
+    */
+   @Named("datacenter:get")
+   @GET   
+   @Path("/{id}")
+   @ResponseParser(DataCenterParser.class)
+   @Produces("application/vnd.profitbricks.resource+json")
+   @Fallback(NullOnNotFoundOr404.class)
+   DataCenter getDataCenter(@PathParam("id") String id);
+
+   @Named("datacenter:get")
+   @GET   
+   @Path("/{id}")
+   @ResponseParser(DataCenterParser.class)
+   @Produces("application/vnd.profitbricks.resource+json")
+   @Fallback(NullOnNotFoundOr404.class)
+   DataCenter getDataCenter(@PathParam("id") String id, DepthOptions options);
+   
+   @Named("datacenter:create")
+   @POST
+   @ResponseParser(DataCenterParser.class)
+   @Produces("application/vnd.profitbricks.resource+json")
+   @MapBinder(DataCenterCreateMapBinder.class)
+   @Fallback(NullOnNotFoundOr404.class)
```

Comment:
Just POST/PUT/PATCH, because the failure is ambiguous. What does a null value mean in that case? The resource does not exist? The operation completed and the modified Object is now `null` (the response did not return a body)? These semantics with null are particularly ambiguous in POST/PUT/PATCH, so we avoid the fallbacks and let the method fail.

In DELETE we can be less restrictive, as it could make sense to return null. The object being deleted is no longer there, so there are no conceptual conflicts.